static/js/tracking.ts

```diff
@@ -0,0 +1,231 @@
+        setTrackingInterval(); // Resume with fresh timer
+    }
+}
+
+function removeTrackingInterval() {
+    const storedData = localStorage.getItem(INTERVAL_KEY);
+    if (storedData) {
+        try {
+            const parsedData = JSON.parse(storedData);
+            clearInterval(parsedData.id); // Clear any potentially lingering timer
+            // console.log(parsedData.id, " interval was removed")
+        } catch (error) {
+            console.error("Error parsing tracking interval data:", error);
+        }
+    }
+
+}
+
+function setTrackingInterval() {
+  const timerId = setInterval(checkUserActivity, WAITING_TIME);
+  localStorage.setItem(INTERVAL_KEY, JSON.stringify({ id: timerId, timestamp: Date.now() }));
+}
+
+
+async function trackEvent(event: Event) {
+    // the following check is necessary since some elements issue click and change events.
+    const currentTime = Date.now();
+    const inactiveDuration = currentTime - lastActiveTime;
+    if (inactiveDuration <= 200) {
+        return;
+    }
+    const target = event.target as HTMLElement;
+    // console.log(target, event.type)
+    if (target.matches('button') || target.matches('a') || target.matches('input') || target.matches('select') || target.matches("div")) {
+        let elementIdOrName = target.id;
+
+        if (!elementIdOrName && target.hasAttribute("name")) {
+            elementIdOrName = target.getAttribute("name") || "";
+        }
+
+        if (ELEMENT_TO_TRACK.includes(elementIdOrName)) {
+            clickCounts = handleLocalStorage(CLICK_COUNTS);
+            
+            const page = window.location.pathname;
+
+            const value = (target as HTMLInputElement).value || "";
+
+            clickCounts.push({time: currentTime, id: elementIdOrName, page, extra: value});
+            
+            // console.log(target, clickCounts)
+            // console.log(`Event: ${event.type}, Element ID or Name: ${elementIdOrName}, Click Count: ${clickCounts}`);
+            handleUserActivity(clickCounts);
+        } 
+    }
+}
+
+// Function to handle user activity
+function handleUserActivity(clickCounts: any) {
+    lastActiveTime = handleLocalStorage(LAST_ACTIVE, Date.now());
+    clickCounts = handleLocalStorage(CLICK_COUNTS, clickCounts);
+    changesSent = false;
+}
+
+// Retrieve or set items in local storage.
+function handleLocalStorage(item: string, value: any = undefined) {
+    const retrievedItem = window.localStorage.getItem(item);
+    if (!retrievedItem || value !== undefined) {
+        value = value || [];
+        window.localStorage.setItem(item, JSON.stringify(value));
+    } else {
+        if (item === CLICK_COUNTS) {
+            value = JSON.parse(retrievedItem);
+        } else if (item === LAST_ACTIVE) {
+            value = parseInt(retrievedItem);
+        }
+    }
+    return value;
+}
+
+// Function to check user activity and send request if inactive for 5 minutes
+async function checkUserActivity() {
```

Comment:
I like that you're batching data up, and not sending individual events! We're doing this to be efficient about our network requests and CPU usage. Any amount of batching on a reasonable time scale will do just fine. I don't think we need to wait for the user definitely not having done anything for 5 minutes; we can send every 5 minutes regardless of when the last activity was. We will get the batching benefit either way -- and by waiting for 5 minutes of inactivity, there is a greater chance that this condition will never be satisfied, because kids are constantly doing something. If a user never sits idle for 5 minutes, we will never get any click tracking for them!

Similarly, I don't think we need to reset our interval when users switch away from the browser window. We already have a test to not send an empty array, so if they switch away and stop using us, we won't send any network requests. And a timer running in the background at an interval of 5 minutes is not going to appreciably affect the CPU usage or battery of the kid's computer.